rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -916,233 +914,142 @@ impl CanisterQueues {
-    /// in `debug_assert!()` checks.
+    /// Handles the timing out or shedding of a message from the pool.
     ///
-    /// Time complexity: O(num_messages).
-    fn calculate_input_queues_stats(
-        canister_queues: &BTreeMap<CanisterId, (InputQueue, OutputQueue)>,
-    ) -> InputQueuesStats {
-        let mut stats = InputQueuesStats::default();
-        let response_count = |msg: &RequestOrResponse| match *msg {
-            RequestOrResponse::Request(_) => 0,
-            RequestOrResponse::Response(_) => 1,
+    /// Records the callback of a shed inbound best-effort response. Releases the
+    /// outbound slot reservation of a shed inbound request. Generates and enqueues
+    /// a reject response if the message was an outbound request. Updates the stats
+    /// for the dropped message and (where applicable) the generated response.
+    ///
+    /// `own_canister_id` and `local_canisters` are required to determine the
+    /// correct input queue schedule to update (when generating a reject response).
+    fn on_message_dropped(
+        &mut self,
+        id: message_pool::Id,
+        msg: &RequestOrResponse,
+        own_canister_id: &CanisterId,
+        local_canisters: &BTreeMap<CanisterId, CanisterState>,
+    ) {
+        use Context::*;
+
+        let context = id.context();
+        let remote = match context {
+            Inbound => msg.sender(),
+            Outbound => msg.receiver(),
+        };
+        let (input_queue, output_queue) = self.canister_queues.get_mut(&remote).unwrap();
+        let (queue, reverse_queue) = match context {
+            Inbound => (input_queue, output_queue),
+            Outbound => (output_queue, input_queue),
         };
-        for (q, _) in canister_queues.values() {
-            stats.message_count += q.len();
-            stats.response_count += q.calculate_stat_sum(response_count);
-            stats.reserved_slots += q.reserved_slots() as isize;
-            stats.size_bytes += q.calculate_size_bytes();
-            stats.cycles += q.cycles_in_queue();
-        }
-        stats
-    }
 
-    /// Computes output queues stats from scratch. Used when deserializing and
-    /// in `debug_assert!()` checks.
-    ///
-    /// Time complexity: O(num_messages).
-    fn calculate_output_queues_stats(
-        canister_queues: &BTreeMap<CanisterId, (InputQueue, OutputQueue)>,
-    ) -> OutputQueuesStats {
-        let mut stats = OutputQueuesStats::default();
-        for (_, q) in canister_queues.values() {
-            stats.message_count += q.num_messages();
-            stats.cycles += q.cycles_in_queue();
+        // Ensure that the first reference in a queue is never stale: if we dropped the
+        // message at the head of a queue, advance to the first non-stale reference.
+        //
+        // Defensive check, reference may have already been popped by an earlier
+        // `on_message_dropped()` call if multiple messages got dropped at once.
```

Comment:
Done.

No duplicate response, this is strictly about popping stale references from the front of the queue. E.g. given an output queue containing request references `[1, 2]`; `1` and `2` expire  as part of the same `time_out_messages()` call; `on_message_dropped(1)` will pop both `1` and `2` and generate a reject response for `1`; on_message_dropped(2)` has nothing left to pop, but will still generate a reject response for `2`.

IOW, `on_message_dropped()` gets called exactly once for every expired / shed message (so we generate exactly one reject response iff it is an outbound request; or release exactly one slot reservation iff it is an inbound request; etc.). But we cannot safely call `queue.peek().unwrap()` on every `on_message_dropped()`, because the queue may have just been emptied.

I guess I intended it as a warning against rewriting the `if` below as  `queue.peek().unwrap()`.